### PR TITLE
User model/testing

### PR DESCRIPTION
**Pull request by rgomezp**
# Description
## 1 Line Summary
Sets up the new generation of test coverage with Jest and adds testing for Core Module functionality.

## Details
* Removes AVA tests from CI and replaces with Jest
* Adds test support directory with testing files
* Adds test coverage for Core Module

# Systems Affected
   - [x] WebSDK
   - [ ] Backend
   - [ ] Dashboard

# Validation
### Info
All new tests. Does **not** include test coverage for the rest of the codebase.

### Checklist
   - [x] All the automated tests pass or I explained why that is not possible
   - [x] I have personally tested this on my machine or explained why that is not possible
   - [x] I have included test coverage for these changes or explained why they are not needed

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Website-SDK/995)

### CODE REVIEW

**Review by jkasten2 — CHANGES_REQUESTED** [review]
Looks like a number of files (the ones in `support/mocks/models/` like `MockPushManager.ts`) are not new in this PR but rather just moved. Can we make a change in this PR so it shows up as moves? Might be able to just delete the old files to get that to show, or might have to do `git mv`

**Review by jkasten2 — CHANGES_REQUESTED** [review]

File: src/sw/serviceWorker/ServiceWorker.ts [hunk 1]
```diff
@@ -101,8 +101,8 @@ export class ServiceWorker {
     self.addEventListener('push', ServiceWorker.onPushReceived);
     self.addEventListener('notificationclose', ServiceWorker.onNotificationClosed);
     self.addEventListener('notificationclick', event => event.waitUntil(ServiceWorker.onNotificationClicked(event)));
-    self.addEventListener('pushsubscriptionchange', (event: PushSubscriptionChangeEvent) => {
-      event.waitUntil(ServiceWorker.onPushSubscriptionChange(event));
+    self.addEventListener('pushsubscriptionchange', (event: Event) => {
+      (event as FetchEvent).waitUntil(ServiceWorker.onPushSubscriptionChange(event as unknown as SubscriptionChangeEvent));
```
[the remark this answers is not in the document]
Comment (rgomezp) [reply src/sw/serviceWorker/ServiceWorker.ts:105]: Typescript removed support for `PushSubscriptionChangeEvent` in 4.4 unfortunately, and putting `SubscriptionChangeEvent` directly as the type for the `event` causes the `addEventListener` function to complain. I was unable to find a replacement and looking directly at `lib.webworker.d.ts`, there is no equivalent that I can see (look for `oldSubscription`, `newSubscription`).

Do you know any alternative solutions?

**Comment by rgomezp** [comment]
> Looks like a number of files (the ones in `support/mocks/models/` like `MockPushManager.ts`) are not new in this PR but rather just moved. Can we make a change in this PR so it shows up as moves? Might be able to just delete the old files to get that to show, or might have to do `git mv`

Done

File: __test__/unit/core/modelRepo.test.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    const emailSubModels = coreDirector.getEmailSubscriptionModels();
+    const emailModel = emailSubModels[newSub.modelId];
+    expect(emailModel.data.token).toBe(TOKEN);
+  });
+
+  /* D E L T A S */
+  test('ModelRepo add subscription -> delta is broadcasted once', (done: jest.DoneCallback) => {
+    core.modelRepo?.subscribe(() => {
+      broadcastCount+=1;
+      passIfBroadcastNTimes(1, broadcastCount, done);
```
[the remark this answers is not in the document]
Comment (rgomezp) [reply __test__/unit/core/modelRepo.test.ts:79]: Good point. Jest will not fail in this case

File: src/sw/serviceWorker/ServiceWorker.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkasten2) [reply src/sw/serviceWorker/ServiceWorker.ts:105]: We have `export interface SubscriptionChangeEvent` in `src/sw/serviceWorker/types.ts` that is breaking this. After removing VSCode shows no errors when doing this now:
```ts
    self.addEventListener('pushsubscriptionchange', (event: SubscriptionChangeEvent) => {
      event.waitUntil(ServiceWorker.onPushSubscriptionChange(event));
    });
```

File: __test__/unit/core/modelRepo.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkasten2) [reply __test__/unit/core/modelRepo.test.ts:79]: The problem with that though is it makes this test fragile (passing when it shouldn't), as if that timeout handling changes in the future this test wouldn't catch that.

File: src/sw/serviceWorker/ServiceWorker.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkasten2) [reply src/sw/serviceWorker/ServiceWorker.ts:105]: It was discovered that Typescript definition for  `pushsubscriptionchange` [is wrong](https://github.com/microsoft/TypeScript/blob/main/lib/lib.webworker.d.ts#L3128) and isn't defined to what it is in the [w3c spec](https://w3c.github.io/push-api/#the-pushsubscriptionchange-event). Given this the code we have here seems to be needed to work around this issue. There isn't a clear way to correct wrong typescript definitions at the project level.

File: __test__/unit/core/modelRepo.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkasten2) [reply __test__/unit/core/modelRepo.test.ts:79]: Didn't understand the `done` until now. This covers what I was talking about above, if `done` is never called the test will fail. What we have here works.

**Review by jkasten2 — APPROVED** [review]

**Pull request merged**